[FSDP][5/N] Add manual "wrapping" support for `fully_shard`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91044 [FSDP][7/N] Support `replicate` in `fully_shard`
* #90959 [FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal
* **#90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`**
* #90871 [FSDP][4/N] Refactor func to share state/init handle attrs
* #90945 fully_shard load state_dict

This PR adds manual "wrapping" support for `fully_shard`. For example, for
```
fully_shard(mod.sub)
fully_shard(mod)
```
`mod.sub` and `mod` will share the same FSDP data structures.

To have parity with wrapper FSDP, this PR only checks support for when each manual application of `fully_shard` passes `policy=None`. Hybrid auto / manual wrapping is not in scope for this PR since it is not supported for wrapper FSDP either. I can follow up to either add support properly or raise and error early.

Commit:
[FSDP][5/N] Add manual "wrapping" support for `fully_shard`

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This is not ready for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This is not ready for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This is not ready for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This is not ready for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This is not ready for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This is not ready for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This is not ready for review.

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/_exec_order_utils.py:
```diff
@@ -366,11 +361,9 @@ def _get_names_from_handles(
             flat_param = handle.flat_param
-            if flat_param not in self.flat_param_to_prefixed_param_names:
+            if flat_param not in self.param_to_fqn:
                 continue
```
when is a handle invalid?

awgu replied:
I traced through the history of my PRs, and it looks like this check was added arbitrarily. However, reasoning about it now, I feel like this check is important for execution to not crash for `use_orig_params=True` since `self.param_to_fqn` is constructed via `_get_param_to_fqns(root_module)` and for `use_orig_params=True`, `_get_param_to_fqns` does not include any `FlatParameter`s since they are not registered.

I think `_exec_order_utils.py` needs to be revisited for `use_orig_params=True`.

mrshenli commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -873,8 +875,7 @@ def _post_backward_final_callback(
 
-    states = [state] if _is_composable(state) else _get_fsdp_states(state)
-    for state in states:
+    for state in _get_fsdp_states(module):
```
hmm, is this intentional to abuse the var name state?

mrshenli replied:
and on line 887, is it a bug? I assume your intention is to set it on the original `state` var instead of the one used in the loop?

awgu replied:
Oh man. Thanks for the great catch. Let me fix this.

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This PR adds manual "wrapping" support for `fully_shard`. For example, for
```
fully_shard(mod.sub)
fully_shard(mod)
```
`mod.sub` and `mod` will share the same FSDP data structures.

To have parity with wrapper FSDP, this PR only checks support for when each manual application of `fully_shard` passes `policy=None`. Hybrid auto / manual wrapping is not in scope for this PR since it is not supported for wrapper FSDP either. I can follow up to either add support properly or raise and error early.




[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This PR adds manual "wrapping" support for `fully_shard`. For example, for
```
fully_shard(mod.sub)
fully_shard(mod)
```
`mod.sub` and `mod` will share the same FSDP data structures.

To have parity with wrapper FSDP, this PR only checks support for when each manual application of `fully_shard` passes `policy=None`. Hybrid auto / manual wrapping is not in scope for this PR since it is not supported for wrapper FSDP either. I can follow up to either add support properly or raise and error early.




[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This PR adds manual "wrapping" support for `fully_shard`. For example, for
```
fully_shard(mod.sub)
fully_shard(mod)
```
`mod.sub` and `mod` will share the same FSDP data structures.

To have parity with wrapper FSDP, this PR only checks support for when each manual application of `fully_shard` passes `policy=None`. Hybrid auto / manual wrapping is not in scope for this PR since it is not supported for wrapper FSDP either. I can follow up to either add support properly or raise and error early.




[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This PR adds manual "wrapping" support for `fully_shard`. For example, for
```
fully_shard(mod.sub)
fully_shard(mod)
```
`mod.sub` and `mod` will share the same FSDP data structures.

To have parity with wrapper FSDP, this PR only checks support for when each manual application of `fully_shard` passes `policy=None`. Hybrid auto / manual wrapping is not in scope for this PR since it is not supported for wrapper FSDP either. I can follow up to either add support properly or raise and error early.




[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Add manual "wrapping" support for `fully_shard`"


This PR adds manual "wrapping" support for `fully_shard`. For example, for
```
fully_shard(mod.sub)
fully_shard(mod)
```
`mod.sub` and `mod` will share the same FSDP data structures.

To have parity with wrapper FSDP, this PR only checks support for when each manual application of `fully_shard` passes `policy=None`. Hybrid auto / manual wrapping is not in scope for this PR since it is not supported for wrapper FSDP either. I can follow up to either add support properly or raise and error early.




[ghstack-poisoned]
